File: Heap_Priority_Queue/1464_Maximum_Product_of_Two_Elements_in_an_Array/1464_Maximum_Product_of_Two_Elements_in_an_Array.java
class Solution {
    public int maxProduct(int[] nums) {

        // 1. Use a max-heap (priority queue) to keep track of largest and second largest elements
        // 2. Pop the two largest elements and return desired product 

        // Use a max heap: negate the values to simulate a max heap
        PriorityQueue<Integer> max_heap = new PriorityQueue<>(nums.length, Collections.reverseOrder());

        for(Integer num : nums)
            max_heap.add(num);

        // pop the two largest elements
        Integer max1 = max_heap.poll();
        Integer max2 = max_heap.poll();

        return (max1-1)*(max2-1);
        
    }
}

File: Heap_Priority_Queue/1464_Maximum_Product_of_Two_Elements_in_an_Array/1464_Maximum_Product_of_Two_Elements_in_an_Array.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {

        // 1. Use a max-heap (priority queue) to keep track of largest and second largest elements
        // 2. Pop the two largest elements and return desired product 

        priority_queue<int> max_heap;

        for(auto& num : nums)
            max_heap.push(num);

        // pop the two largest elements
        int max1 = max_heap.top();
        max_heap.pop();
        int max2 = max_heap.top();

        return (max1-1)*(max2-1);

    }
};